<a href="https://colab.research.google.com/github/phurisk/BSC_DPDM23/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [258]:
import pandas as pd
import os

In [259]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23/File'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [260]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [261]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

## แปลง data ให้อยู่ในรูปแบบ transaction data ด้วย group by .appy(list)

In [262]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [263]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

- ดูประเทศ

In [264]:
set(data["Country"])

{'Australia',
 'Austria',
 'Bahrain',
 'Belgium',
 'Brazil',
 'Canada',
 'Channel Islands',
 'Cyprus',
 'Czech Republic',
 'Denmark',
 'EIRE',
 'European Community',
 'Finland',
 'France',
 'Germany',
 'Greece',
 'Hong Kong',
 'Iceland',
 'Israel',
 'Italy',
 'Japan',
 'Lebanon',
 'Lithuania',
 'Malta',
 'Netherlands',
 'Norway',
 'Poland',
 'Portugal',
 'RSA',
 'Saudi Arabia',
 'Singapore',
 'Spain',
 'Sweden',
 'Switzerland',
 'USA',
 'United Arab Emirates',
 'United Kingdom',
 'Unspecified'}

## Association Rules

In [265]:
!pip install apyori

In [266]:
from apyori import apriori

In [267]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [268]:
len(frequentItemsets)

4

In [269]:
frequentItemsets[0][0]

frozenset({'22423'})

In [270]:
# dir(frequentItemsets[0][0])

In [271]:
frequentItemsets[0][0].__len__()

1

In [272]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [273]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [274]:
# set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

In [275]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country  
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom

## quiz 22386 และ 85099b คือสินค้าอะไร

In [276]:
set(data[(data['StockCode']=='22386')  ]['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [277]:
set(data[(data['StockCode']=='85099B') ]['Description'])

{'JUMBO BAG RED RETROSPOT'}

#Hw หา Frequent Item Set  และ Association Rules
- ของช่วงเวลา Chirtmas & New Year(1 Dec -31 Dec)
- ของประเทศในทวีป Europe
- ของช่วงเวลา ตรุษจีน ในทวีป Asai (15 jan-15 Feb)
- ของช่วงเวลา Chirtmas & New Year ในทวีป Europe (1 Dec -31 Dec)
- ของวันวาเลนไทน์
- โดยจัดกลุ่มตามประเทศ+เดือน

In [278]:
!pip install apyori

In [279]:
from apyori import apriori

In [280]:
df = pd.read_csv(os.path.join(path, 'reduced_marketbasket.csv'), encoding='latin-1')


## ข้อ 1. ช่วงเวลา Chirtmas & New Year(1 Dec -31 Dec)

In [281]:
# Convert รูปแบบเวลาใน column InvoiceDate ให้อยู่ในรูปที่ต้องการ
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'], format='%m/%d/%Y %H:%M')


In [282]:
# Filter December 1st to December 31st
filtered_df = df[(df['InvoiceDate'].dt.month == 12) & (df['InvoiceDate'].dt.day >= 1) & (df['InvoiceDate'].dt.day <= 31)]

#ปริ้นเพื่อตรวจสอบ
print(filtered_df)

       InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      

In [283]:
# transactions= filtered_df.groupby('InvoiceNo')['StockCode'].apply(list)
# transactions  ### ถ้าใช้ transactions แทน transactions_apply จะเป็นการโชว์ StockCode แทนข้อมูลสินค้า ###

In [ ]:
transactions = filtered_df.groupby('InvoiceNo')['Description'].apply(list).reset_index() #ทำการ group และทำให้ข้อมูลใน Description เป็นlist
transactions_apply = transactions['Description'].apply(lambda x: [str(item) for item in x]) #ใช้ฟังก์ชัน lambda เพื่อแปลงสินค้าทุกชิ้นใน list ของ 'transactions' เป็นข้อความ (string)

In [ ]:
frequentItemset = list(apriori(transactions_apply, min_support=0.035))

for record in frequentItemset:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: ALARM CLOCK BAKELIKE RED , ALARM CLOCK BAKELIKE GREEN
Support: 0.0368 (3.68%)
Rule 1: (None -> ALARM CLOCK BAKELIKE RED , ALARM CLOCK BAKELIKE GREEN : Confidence=0.0368, Lift=1.0000)
Rule 2: (ALARM CLOCK BAKELIKE GREEN -> ALARM CLOCK BAKELIKE RED  : Confidence=0.7089, Lift=14.4627)
Rule 3: (ALARM CLOCK BAKELIKE RED  -> ALARM CLOCK BAKELIKE GREEN : Confidence=0.7517, Lift=14.4627)

yes
Items: HOT WATER BOTTLE I AM SO POORLY, CHOCOLATE HOT WATER BOTTLE
Support: 0.0365 (3.65%)
Rule 1: (None -> HOT WATER BOTTLE I AM SO POORLY, CHOCOLATE HOT WATER BOTTLE : Confidence=0.0365, Lift=1.0000)
Rule 2: (CHOCOLATE HOT WATER BOTTLE -> HOT WATER BOTTLE I AM SO POORLY : Confidence=0.5163, Lift=9.7484)
Rule 3: (HOT WATER BOTTLE I AM SO POORLY -> CHOCOLATE HOT WATER BOTTLE : Confidence=0.6894, Lift=9.7484)

yes
Items: PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN KIT VINTAGE CHRISTMAS
Support: 0.0477 (4.77%)
Rule 1: (None -> PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN KIT VINTAGE CHRISTMAS : 

สรุป เลือกอันที่มีค่าสนับสนุนมากที่สุด


Items: PAPER CHAIN KIT VINTAGE CHRISTMAS และ PAPER CHAIN KIT 50'S CHRISTMAS ค่าสนับสนุนคือ 0.0477 หรือ 4.77% ของการทำธุรกรรมทั้งหมด

- เมื่อ 'PAPER CHAIN KIT 50'S CHRISTMAS' ถูกซื้อ มีความมั่นใจ 51.60% ว่า 'PAPER CHAIN KIT VINTAGE CHRISTMAS' จะถูกซื้อด้วย

- เมื่อ 'PAPER CHAIN KIT VINTAGE CHRISTMAS' ถูกซื้อ มีโอกาส 66.51% ว่า 'PAPER CHAIN KIT 50'S CHRISTMAS' จะถูกซื้อด้วย

## ข้อ 2. ประเทศในทวีป Europe

In [286]:
europe_countries = {
    'Austria', 'Belgium', 'Channel Islands', 'Cyprus', 'Czech Republic', 'Denmark', 'EIRE',
    'European Community', 'Finland', 'France', 'Germany', 'Greece', 'Iceland', 'Italy',
    'Lithuania', 'Malta', 'Netherlands', 'Norway', 'Poland', 'Portugal', 'Spain', 'Sweden',
    'Switzerland', 'United Kingdom', 'Unspecified'
}

In [287]:
# เลือกประเทศจาก european_countries
filtered_df2 = df[df['Country'].isin(europe_countries)]

# โชว์ datafram เพื่อcheck
print(filtered_df2)

       InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      

In [288]:
transactions2 = filtered_df2.groupby('InvoiceNo')['Description'].apply(list).reset_index() #ทำการ group และทำให้ข้อมูลใน Description เป็นlist
transactions_apply2 = transactions2['Description'].apply(lambda x: [str(item) for item in x]) #ใช้ฟังก์ชัน lambda เพื่อแปลงสินค้าทุกชิ้นใน list ของ 'transactions' เป็นข้อความ (string)

In [289]:
frequentItemset2 = list(apriori(transactions_apply2, min_support=0.025))

for record in frequentItemset2:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: GREEN REGENCY TEACUP AND SAUCER, ROSES REGENCY TEACUP AND SAUCER 
Support: 0.0301 (3.01%)
Rule 1: (None -> GREEN REGENCY TEACUP AND SAUCER, ROSES REGENCY TEACUP AND SAUCER  : Confidence=0.0301, Lift=1.0000)
Rule 2: (GREEN REGENCY TEACUP AND SAUCER -> ROSES REGENCY TEACUP AND SAUCER  : Confidence=0.7402, Lift=17.2141)
Rule 3: (ROSES REGENCY TEACUP AND SAUCER  -> GREEN REGENCY TEACUP AND SAUCER : Confidence=0.7001, Lift=17.2141)

yes
Items: JUMBO BAG RED RETROSPOT, JUMBO BAG PINK POLKADOT
Support: 0.0323 (3.23%)
Rule 1: (None -> JUMBO BAG RED RETROSPOT, JUMBO BAG PINK POLKADOT : Confidence=0.0323, Lift=1.0000)
Rule 2: (JUMBO BAG PINK POLKADOT -> JUMBO BAG RED RETROSPOT : Confidence=0.6775, Lift=8.2081)
Rule 3: (JUMBO BAG RED RETROSPOT -> JUMBO BAG PINK POLKADOT : Confidence=0.3915, Lift=8.2081)

yes
Items: JUMBO SHOPPER VINTAGE RED PAISLEY, JUMBO BAG RED RETROSPOT
Support: 0.0265 (2.65%)
Rule 1: (None -> JUMBO SHOPPER VINTAGE RED PAISLEY, JUMBO BAG RED RETROSPOT : Confidence=0

## ข้อ 3. ช่วงเวลา ตรุษจีน ในทวีป Asai (15 jan-15 Feb)

In [290]:
asia_countries= {
    'Bahrain', 'Hong Kong', 'Israel', 'Japan', 'Lebanon', 'Saudi Arabia', 'Singapore', 'United Arab Emirates'
}

In [291]:
# เลือกวันตรุษจีน ในทวีป Asai (15 jan-15 Feb)
filtered_df3 = df[
     (((df['InvoiceDate'].dt.month == 1) & (df['InvoiceDate'].dt.day >= 15)) |
    ((df['InvoiceDate'].dt.month == 2) & (df['InvoiceDate'].dt.day <= 15))) &
    (df['Country'].isin(asia_countries))
]


# โชว์ datafram เพื่อcheck
print(filtered_df3)

      InvoiceNo StockCode                          Description  Quantity  \
69623    541975     22620          4 TRADITIONAL SPINNING TOPS        60   
69624    541975    85183B  CHARLIE & LOLA WASTEPAPER BIN FLORA         6   
69625    541975    85183A   CHARLIE & LOLA WASTEPAPER BIN BLUE         6   
69626    541975     16236                 KITTY PENCIL ERASERS        28   
69627    541975     21058               PARTY INVITES WOODLAND        12   
...         ...       ...                                  ...       ...   
89595    543911    47590B          PINK HAPPY BIRTHDAY BUNTING        12   
89596    543911     22117     METAL SIGN HER DINNER IS SERVED         12   
89597    543911     82567            AIRLINE LOUNGE,METAL SIGN        12   
89598    543911     21910                  WAY OUT METAL SIGN         12   
89599    543911     21911                   GARDEN METAL SIGN         12   

              InvoiceDate  UnitPrice  CustomerID               Country  
69623 2011-01-

In [292]:
transactions3 = filtered_df3.groupby('InvoiceNo')['Description'].apply(list).reset_index() #ทำการ group และทำให้ข้อมูลใน Description เป็นlist
transactions_apply3 = transactions3['Description'].apply(lambda x: [str(item) for item in x]) #ใช้ฟังก์ชัน lambda เพื่อแปลงสินค้าทุกชิ้นใน list ของ 'transactions' เป็นข้อความ (string)

In [293]:
frequentItemset3 = list(apriori(transactions_apply3, min_support=0.3))

for record in frequentItemset3:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: BASKET OF TOADSTOOLS, LUNCH BAG RED RETROSPOT
Support: 0.3333 (33.33%)
Rule 1: (None -> BASKET OF TOADSTOOLS, LUNCH BAG RED RETROSPOT : Confidence=0.3333, Lift=1.0000)
Rule 2: (BASKET OF TOADSTOOLS -> LUNCH BAG RED RETROSPOT : Confidence=1.0000, Lift=3.0000)
Rule 3: (LUNCH BAG RED RETROSPOT -> BASKET OF TOADSTOOLS : Confidence=1.0000, Lift=3.0000)

yes
Items: BASKET OF TOADSTOOLS, SET 3 RETROSPOT TEA,COFFEE,SUGAR
Support: 0.3333 (33.33%)
Rule 1: (None -> BASKET OF TOADSTOOLS, SET 3 RETROSPOT TEA,COFFEE,SUGAR : Confidence=0.3333, Lift=1.0000)
Rule 2: (BASKET OF TOADSTOOLS -> SET 3 RETROSPOT TEA,COFFEE,SUGAR : Confidence=1.0000, Lift=3.0000)
Rule 3: (SET 3 RETROSPOT TEA,COFFEE,SUGAR -> BASKET OF TOADSTOOLS : Confidence=1.0000, Lift=3.0000)

yes
Items: JAM MAKING SET WITH JARS, BATH BUILDING BLOCK WORD
Support: 0.3333 (33.33%)
Rule 1: (None -> JAM MAKING SET WITH JARS, BATH BUILDING BLOCK WORD : Confidence=0.3333, Lift=1.0000)
Rule 2: (BATH BUILDING BLOCK WORD -> JAM MAKING SET

## ข้อ 4. ช่วงเวลา Chirtmas & New Year ในทวีป Europe (1 Dec -31 Dec)

In [340]:
# เลือกช่วงเวลา Chirtmas & New Year ในทวีป Europe (1 Dec -31 Dec)
filtered_df4 = df[
     (df['InvoiceDate'].dt.month == 12)  &
    (df['Country'].isin(europe_countries))
]


# โชว์ datafram เพื่อcheck
print(filtered_df4)

       InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      

In [295]:
transactions4 = filtered_df4.groupby('InvoiceNo')['Description'].apply(list).reset_index() #ทำการ group และทำให้ข้อมูลใน Description เป็นlist
transactions_apply4 = transactions4['Description'].apply(lambda x: [str(item) for item in x]) #ใช้ฟังก์ชัน lambda เพื่อแปลงสินค้าทุกชิ้นใน list ของ 'transactions' เป็นข้อความ (string)

In [296]:
frequentItemset4 = list(apriori(transactions_apply4, min_support=0.035))

for record in frequentItemset4:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: ALARM CLOCK BAKELIKE RED , ALARM CLOCK BAKELIKE GREEN
Support: 0.0367 (3.67%)
Rule 1: (None -> ALARM CLOCK BAKELIKE RED , ALARM CLOCK BAKELIKE GREEN : Confidence=0.0367, Lift=1.0000)
Rule 2: (ALARM CLOCK BAKELIKE GREEN -> ALARM CLOCK BAKELIKE RED  : Confidence=0.7070, Lift=14.4602)
Rule 3: (ALARM CLOCK BAKELIKE RED  -> ALARM CLOCK BAKELIKE GREEN : Confidence=0.7500, Lift=14.4602)

yes
Items: HOT WATER BOTTLE I AM SO POORLY, CHOCOLATE HOT WATER BOTTLE
Support: 0.0367 (3.67%)
Rule 1: (None -> HOT WATER BOTTLE I AM SO POORLY, CHOCOLATE HOT WATER BOTTLE : Confidence=0.0367, Lift=1.0000)
Rule 2: (CHOCOLATE HOT WATER BOTTLE -> HOT WATER BOTTLE I AM SO POORLY : Confidence=0.5163, Lift=9.7067)
Rule 3: (HOT WATER BOTTLE I AM SO POORLY -> CHOCOLATE HOT WATER BOTTLE : Confidence=0.6894, Lift=9.7067)

yes
Items: PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN KIT VINTAGE CHRISTMAS
Support: 0.0479 (4.79%)
Rule 1: (None -> PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN KIT VINTAGE CHRISTMAS : 

## ข้อ 5. วันวาเลนไทน์

In [297]:
# เลือกเฉพาะวันวาเลนไทน์
filtered_df5 = df[(df['InvoiceDate'].dt.month == 2) & (df['InvoiceDate'].dt.day == 14) ]

# โชว์ datafram เพื่อcheck
print(filtered_df5)

      InvoiceNo StockCode                         Description  Quantity  \
88847    543821     22926      IVORY GIANT GARDEN THERMOMETER         2   
88848    543821     22928     YELLOW GIANT GARDEN THERMOMETER         2   
88849    543821     22925       BLUE GIANT GARDEN THERMOMETER         2   
88850    543821     85175               CACTI T-LIGHT CANDLES        16   
88851    543821     22456     NATURAL SLATE CHALKBOARD LARGE          3   
...         ...       ...                                 ...       ...   
89950    543979     21889              WOODEN BOX OF DOMINOES         1   
89951    543979     85176          SEWING SUSAN 21 NEEDLE SET         1   
89952    543979     37446  MINI CAKE STAND WITH HANGING CAKES         1   
89953    543979     21657     MILK BOTTLE WITH GLASS STOPPER          1   
89954   C543980     22728           ALARM CLOCK BAKELIKE PINK        -1   

              InvoiceDate  UnitPrice  CustomerID         Country  
88847 2011-02-14 09:24:00       

In [298]:
transactions5 = filtered_df5.groupby('InvoiceNo')['Description'].apply(list).reset_index() #ทำการ group และทำให้ข้อมูลใน Description เป็นlist
transactions_apply5 = transactions5['Description'].apply(lambda x: [str(item) for item in x]) #ใช้ฟังก์ชัน lambda เพื่อแปลงสินค้าทุกชิ้นใน list ของ 'transactions' เป็นข้อความ (string)

In [299]:
frequentItemset4 = list(apriori(transactions_apply4, min_support=0.035))

for record in frequentItemset4:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: ALARM CLOCK BAKELIKE RED , ALARM CLOCK BAKELIKE GREEN
Support: 0.0367 (3.67%)
Rule 1: (None -> ALARM CLOCK BAKELIKE RED , ALARM CLOCK BAKELIKE GREEN : Confidence=0.0367, Lift=1.0000)
Rule 2: (ALARM CLOCK BAKELIKE GREEN -> ALARM CLOCK BAKELIKE RED  : Confidence=0.7070, Lift=14.4602)
Rule 3: (ALARM CLOCK BAKELIKE RED  -> ALARM CLOCK BAKELIKE GREEN : Confidence=0.7500, Lift=14.4602)

yes
Items: HOT WATER BOTTLE I AM SO POORLY, CHOCOLATE HOT WATER BOTTLE
Support: 0.0367 (3.67%)
Rule 1: (None -> HOT WATER BOTTLE I AM SO POORLY, CHOCOLATE HOT WATER BOTTLE : Confidence=0.0367, Lift=1.0000)
Rule 2: (CHOCOLATE HOT WATER BOTTLE -> HOT WATER BOTTLE I AM SO POORLY : Confidence=0.5163, Lift=9.7067)
Rule 3: (HOT WATER BOTTLE I AM SO POORLY -> CHOCOLATE HOT WATER BOTTLE : Confidence=0.6894, Lift=9.7067)

yes
Items: PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN KIT VINTAGE CHRISTMAS
Support: 0.0479 (4.79%)
Rule 1: (None -> PAPER CHAIN KIT 50'S CHRISTMAS , PAPER CHAIN KIT VINTAGE CHRISTMAS : 

## ข้อ 6. โดยจัดกลุ่มตามประเทศ+เดือน

In [300]:
df6 = pd.read_csv(os.path.join(path, 'reduced_marketbasket.csv'), encoding='latin-1')


In [312]:
# Convert รูปแบบเวลาใน column InvoiceDate ให้อยู่ในรูปที่ต้องการ
df6['InvoiceDate'] = pd.to_datetime(df6['InvoiceDate'], format='%m/%d/%Y %H:%M')

In [313]:
df6['Month'] = df6['InvoiceDate'].dt.month

# df6['Year'] = df6['InvoiceDate'].dt.year   #ถ้าต้องการ"ปี"ด้วย

In [303]:
# transactions_Mon_Con = df6.groupby(['Country', 'Month'])['StockCode'].apply(list)

In [314]:
transactions6 = df6.groupby(['Country', 'Month'])['StockCode'].apply(list) #ทำการ group และทำให้ข้อมูลใน Description เป็นlist  สามารถเพิ่มปี ได้โดย 'Year'
# transactions_apply6 = transactions6['Description'].apply(lambda x: [str(item) for item in x]) #ใช้ฟังก์ชัน lambda เพื่อแปลงสินค้าทุกชิ้นใน list ของ 'transactions' เป็นข้อความ (string


In [315]:
frequentItemset6 = list(apriori(transactions6, min_support=0.30))

for record in frequentItemset6:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: POST, 22326
Support: 0.3209 (32.09%)
Rule 1: (None -> POST, 22326 : Confidence=0.3209, Lift=1.0000)
Rule 2: (22326 -> POST : Confidence=0.8407, Lift=1.4638)
Rule 3: (POST -> 22326 : Confidence=0.5588, Lift=1.4638)

yes
Items: 22423, POST
Support: 0.3142 (31.42%)
Rule 1: (None -> 22423, POST : Confidence=0.3142, Lift=1.0000)
Rule 2: (22423 -> POST : Confidence=0.6838, Lift=1.1907)
Rule 3: (POST -> 22423 : Confidence=0.5471, Lift=1.1907)

yes
Items: 22630, 22629
Support: 0.3007 (30.07%)
Rule 1: (None -> 22630, 22629 : Confidence=0.3007, Lift=1.0000)
Rule 2: (22629 -> 22630 : Confidence=0.8812, Lift=2.5572)
Rule 3: (22630 -> 22629 : Confidence=0.8725, Lift=2.5572)



### 6.1 ช่วงเวลา Chirtmas & New Year(1 Dec -31 Dec)

In [321]:
# Filter December 1st to December 31st
filtered_df61 = df6[(df6['InvoiceDate'].dt.month == 12) & (df6['InvoiceDate'].dt.day >= 1) & (df6['InvoiceDate'].dt.day <= 31)]


In [322]:
transactions61 = filtered_df61.groupby(['Country', 'Month'])['StockCode'].apply(list)

In [323]:
frequentItemset61 = list(apriori(transactions61, min_support=0.3))

for record in frequentItemset61:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: POST, 20750
Support: 0.3333 (33.33%)
Rule 1: (None -> POST, 20750 : Confidence=0.3333, Lift=1.0000)
Rule 2: (20750 -> POST : Confidence=0.9000, Lift=1.7357)
Rule 3: (POST -> 20750 : Confidence=0.6429, Lift=1.7357)

yes
Items: 22727, 22728
Support: 0.3333 (33.33%)
Rule 1: (None -> 22727, 22728 : Confidence=0.3333, Lift=1.0000)
Rule 2: (22727 -> 22728 : Confidence=1.0000, Lift=3.0000)
Rule 3: (22728 -> 22727 : Confidence=1.0000, Lift=3.0000)



### 6.2 ช่วงเวลา Chirtmas & New Year(1 Dec -31 Dec)

In [331]:
# เลือกประเทศจาก european_countries
filtered_df62 = df6[df6['Country'].isin(europe_countries)]

In [332]:
transactions62 = filtered_df62.groupby(['Country', 'Month'])['StockCode'].apply(list)

In [333]:
frequentItemset62 = list(apriori(transactions62, min_support=0.35))

for record in frequentItemset61:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: 22423, 21212
Support: 0.3519 (35.19%)
Rule 1: (None -> 22423, 21212 : Confidence=0.3519, Lift=1.0000)
Rule 2: (21212 -> 22423 : Confidence=0.7885, Lift=1.5702)
Rule 3: (22423 -> 21212 : Confidence=0.7009, Lift=1.5702)

yes
Items: POST, 21212
Support: 0.3605 (36.05%)
Rule 1: (None -> POST, 21212 : Confidence=0.3605, Lift=1.0000)
Rule 2: (21212 -> POST : Confidence=0.8077, Lift=1.1475)
Rule 3: (POST -> 21212 : Confidence=0.5122, Lift=1.1475)

yes
Items: POST, 22326
Support: 0.3991 (39.91%)
Rule 1: (None -> POST, 22326 : Confidence=0.3991, Lift=1.0000)
Rule 2: (22326 -> POST : Confidence=0.9118, Lift=1.2954)
Rule 3: (POST -> 22326 : Confidence=0.5671, Lift=1.2954)

yes
Items: 22423, POST
Support: 0.3863 (38.63%)
Rule 1: (None -> 22423, POST : Confidence=0.3863, Lift=1.0000)
Rule 2: (22423 -> POST : Confidence=0.7692, Lift=1.0929)
Rule 3: (POST -> 22423 : Confidence=0.5488, Lift=1.0929)

yes
Items: POST, 22554
Support: 0.3691 (36.91%)
Rule 1: (None -> POST, 22554 : Confidence=0.

### 6.3 ช่วงเวลา Chirtmas & New Year(1 Dec -31 Dec)

In [337]:
filtered_df63 = df6[
     (((df6['InvoiceDate'].dt.month == 1) & (df6['InvoiceDate'].dt.day >= 15)) |
    ((df6['InvoiceDate'].dt.month == 2) & (df6['InvoiceDate'].dt.day <= 15))) &
    (df6['Country'].isin(asia_countries))
]

In [338]:
transactions63 = filtered_df63.groupby(['Country', 'Month'])['StockCode'].apply(list)

In [339]:
frequentItemset63 = list(apriori(transactions63, min_support=0.35))

for record in frequentItemset61:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: 22423, 21212
Support: 0.3519 (35.19%)
Rule 1: (None -> 22423, 21212 : Confidence=0.3519, Lift=1.0000)
Rule 2: (21212 -> 22423 : Confidence=0.7885, Lift=1.5702)
Rule 3: (22423 -> 21212 : Confidence=0.7009, Lift=1.5702)

yes
Items: POST, 21212
Support: 0.3605 (36.05%)
Rule 1: (None -> POST, 21212 : Confidence=0.3605, Lift=1.0000)
Rule 2: (21212 -> POST : Confidence=0.8077, Lift=1.1475)
Rule 3: (POST -> 21212 : Confidence=0.5122, Lift=1.1475)

yes
Items: POST, 22326
Support: 0.3991 (39.91%)
Rule 1: (None -> POST, 22326 : Confidence=0.3991, Lift=1.0000)
Rule 2: (22326 -> POST : Confidence=0.9118, Lift=1.2954)
Rule 3: (POST -> 22326 : Confidence=0.5671, Lift=1.2954)

yes
Items: 22423, POST
Support: 0.3863 (38.63%)
Rule 1: (None -> 22423, POST : Confidence=0.3863, Lift=1.0000)
Rule 2: (22423 -> POST : Confidence=0.7692, Lift=1.0929)
Rule 3: (POST -> 22423 : Confidence=0.5488, Lift=1.0929)

yes
Items: POST, 22554
Support: 0.3691 (36.91%)
Rule 1: (None -> POST, 22554 : Confidence=0.

### 6.4 ช่วงเวลา Chirtmas & New Year(1 Dec -31 Dec)

In [346]:
filtered_df64 = df6[
     (df6['InvoiceDate'].dt.month == 12)  &
    (df6['Country'].isin(europe_countries))
]

In [347]:
transactions64 = filtered_df64.groupby(['Country', 'Month'])['StockCode'].apply(list)

In [348]:
frequentItemset64 = list(apriori(transactions64, min_support=0.35))

for record in frequentItemset61:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: 22423, 21212
Support: 0.3519 (35.19%)
Rule 1: (None -> 22423, 21212 : Confidence=0.3519, Lift=1.0000)
Rule 2: (21212 -> 22423 : Confidence=0.7885, Lift=1.5702)
Rule 3: (22423 -> 21212 : Confidence=0.7009, Lift=1.5702)

yes
Items: POST, 21212
Support: 0.3605 (36.05%)
Rule 1: (None -> POST, 21212 : Confidence=0.3605, Lift=1.0000)
Rule 2: (21212 -> POST : Confidence=0.8077, Lift=1.1475)
Rule 3: (POST -> 21212 : Confidence=0.5122, Lift=1.1475)

yes
Items: POST, 22326
Support: 0.3991 (39.91%)
Rule 1: (None -> POST, 22326 : Confidence=0.3991, Lift=1.0000)
Rule 2: (22326 -> POST : Confidence=0.9118, Lift=1.2954)
Rule 3: (POST -> 22326 : Confidence=0.5671, Lift=1.2954)

yes
Items: 22423, POST
Support: 0.3863 (38.63%)
Rule 1: (None -> 22423, POST : Confidence=0.3863, Lift=1.0000)
Rule 2: (22423 -> POST : Confidence=0.7692, Lift=1.0929)
Rule 3: (POST -> 22423 : Confidence=0.5488, Lift=1.0929)

yes
Items: POST, 22554
Support: 0.3691 (36.91%)
Rule 1: (None -> POST, 22554 : Confidence=0.

### 6.5 ช่วงเวลา Chirtmas & New Year(1 Dec -31 Dec)

In [349]:
filtered_df65 = df6[(df6['InvoiceDate'].dt.month == 2) & (df6['InvoiceDate'].dt.day == 14) ]

In [350]:
transactions65 = filtered_df65.groupby(['Country', 'Month'])['StockCode'].apply(list)

In [351]:
frequentItemset65 = list(apriori(transactions65, min_support=0.35))

for record in frequentItemset61:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: 22423, 21212
Support: 0.3519 (35.19%)
Rule 1: (None -> 22423, 21212 : Confidence=0.3519, Lift=1.0000)
Rule 2: (21212 -> 22423 : Confidence=0.7885, Lift=1.5702)
Rule 3: (22423 -> 21212 : Confidence=0.7009, Lift=1.5702)

yes
Items: POST, 21212
Support: 0.3605 (36.05%)
Rule 1: (None -> POST, 21212 : Confidence=0.3605, Lift=1.0000)
Rule 2: (21212 -> POST : Confidence=0.8077, Lift=1.1475)
Rule 3: (POST -> 21212 : Confidence=0.5122, Lift=1.1475)

yes
Items: POST, 22326
Support: 0.3991 (39.91%)
Rule 1: (None -> POST, 22326 : Confidence=0.3991, Lift=1.0000)
Rule 2: (22326 -> POST : Confidence=0.9118, Lift=1.2954)
Rule 3: (POST -> 22326 : Confidence=0.5671, Lift=1.2954)

yes
Items: 22423, POST
Support: 0.3863 (38.63%)
Rule 1: (None -> 22423, POST : Confidence=0.3863, Lift=1.0000)
Rule 2: (22423 -> POST : Confidence=0.7692, Lift=1.0929)
Rule 3: (POST -> 22423 : Confidence=0.5488, Lift=1.0929)

yes
Items: POST, 22554
Support: 0.3691 (36.91%)
Rule 1: (None -> POST, 22554 : Confidence=0.